src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/AutoClosingTagOnAutoInsertProvider.cs

```diff
@@ -209,6 +209,23 @@ private static bool TryEnsureOwner_WorkaroundCompilerQuirks(int afterCloseAngleI
                 var closeAngleSoureChange = new SourceChange(closeAngleIndex, length: 0, newText: string.Empty);
                 currentOwner = syntaxTree.Root.LocateOwner(closeAngleSoureChange);
             }
+            else if (currentOwner.Parent is MarkupEndTagSyntax ||
+                     currentOwner.Parent is MarkupTagHelperEndTagSyntax)
+            {
+                // Quirk: https://github.com/dotnet/aspnetcore/issues/33919#issuecomment-870233627
+                // When tags are nested within each other within a C# block like:
+                //
+                // @if (true)
+                // {
+                //     <div><em>|</div>
+                // }
+                //
+                // The owner will be the </div>. Note this does not happen outside of C# blocks.
+
+                var closeAngleIndex = afterCloseAngleIndex - 1;
```

Comment:
nit: Since this code seems to be identical to the case above, maybe extracting this into a local helper function could be good?

Reply:
You're right in that it's the same; however in this scenario I choose to duplicate the code because conceptually the two are different (one incorrectly matches an end-tag and another matches C#). Due to that conceptual difference I like to keep logic separated so changing one doesn't inadvertently break another.